proposals/0000-updated-integers.md

```diff
@@ -0,0 +1,718 @@
+  /// returns `nil` otherwise.
+  init?<T : Integer>(exactly source: T)
+
+  func adding( rhs: Self) -> Self
+  func subtracting( rhs: Self) -> Self
+  func multiplied(by rhs: Self) -> Self
+  func divided(by rhs: Self) -> Self
+
+  mutating func add( rhs: Self)
+  mutating func subtract( rhs: Self)
+  mutating func multiply(by rhs: Self)
+  mutating func divide(by rhs: Self)
+}
+
+extension Arithmetic {
+  public init() { self = 0 }
+}
+```
+
+#### `SignedArithmetic`
+
+The `SignedArithmetic` protocol is for numbers that can be negated.
+
+```Swift
+public protocol SignedArithmetic : Arithmetic {
+  func negated() -> Self
+  mutating func negate()
+}
+
+extension SignedArithmetic {
+  public func negated() -> Self {
+    return Self() - self
+  }
+  public mutating func negate() {
+    self = negated()
+  }
+}
+```
+
+#### `Integer`
+
+The `Integer` protocol is the basis for all the integer types provided by the
+standard library.
+
+The `isEqual(to:)` and `isLess(than:)` methods provide implementations for
+`Equatable` and `Comparable` protocol conformances. Similar to how arithmetic
+operations are dispatched in `Arithmetic`, `==` and `<` operators for
+homogeneous comparisons are implemented as generic free functions invoking the
+`isEqual(to:)` and `isLess(than:)` protocol methods respectively.
+
+This protocol adds 4 new initializers. One of them allows to create integers
+from floating point numbers, if the value is representable exactly, others
+support construction from instances of any type conforming to `Integer`, using
+different strategies:
+
+  - Initialze `Self` with the value, provided that the value is representable.
+    The precondition should be satisfied by the caller.
+
+  - Extend or truncate the value to fit into `Self`
+
+  - Clamp the value to the representable range of `Self`
+
+The `AbsoluteValue` associated type is able to hold the absolute value of any
+possible value of `Self`. Concrete types do not have to provide a typealias for
+it, as it can be inferred from the `absoluteValue` property. This property can
+be useful in operations that are simpler to implement in terms of unsigned
+values, for example, printing a value of an integer, which is just printing a
+'-' character in front of an absolute value.
+
+Please note that for ordinary work, the `absoluteValue` property **should not**
+be preferred to the `abs(_)` function, whose return value is of the same type
+as its argument.
+
+```Swift
+public protocol Integer:
+  Comparable, Arithmetic,
+  IntegerLiteralConvertible, CustomStringConvertible {
+
+  associatedtype AbsoluteValue : Integer // this is not the actual code
+
+  static var isSigned: Bool { get }
+
+  var absoluteValue: AbsoluteValue { get }
+
+  func isEqual(to rhs: Self) -> Bool
+  func isLess(than rhs: Self) -> Bool
+
+  /// Creates an instance of `Self` that has the exact value of `source`,
+  /// returns `nil` otherwise.
+  init?<T : FloatingPoint>(exactly source: T)
```

Comment:
It makes sense to have the truncating init as well; we already have it for all concrete stdlib integer and floating-point types as `init(_ source: ${FloatType})`, so this would just be putting it in the protocol.

For other rounding directions, rounding and then converting makes sense, though we'll need to add `ceil`, `floor`, etc to the `FloatingPoint` protocol to make that work smoothly (we should do that anyway).